sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1445,45 +1494,108 @@ _pi_queue::pi_queue_group_t::getZeQueue(uint32_t *QueueGroupOrdinal) {
+  if (Index != 0) {
     ZeCommandQueueDesc.flags = ZE_COMMAND_QUEUE_FLAG_EXPLICIT_ONLY;
   }
 
   zePrint("[getZeQueue]: create queue ordinal = %d, index = %d "
           "(round robin in [%d, %d])\n",
           ZeCommandQueueDesc.ordinal, ZeCommandQueueDesc.index, LowerIndex,
           UpperIndex);
+}
+
+// This function will return one of possibly multiple available native
+// queues and the value of the queue group ordinal.
+ze_command_queue_handle_t &
+_pi_queue::pi_queue_group_t::getZeQueue(uint32_t *QueueGroupOrdinal) {
+
+  // QueueIndex is the proper L0 index.
+  // Index is the plugins concept of index, with main and link copy engines in
+  // one range.
+  uint32_t QueueIndex;
+  auto Index = getQueueIndex(QueueGroupOrdinal, &QueueIndex);
+
+  ze_command_queue_handle_t &ZeQueue = ZeQueues[Index];
+  if (ZeQueue)
+    return ZeQueue;
+
+  ZeStruct<ze_command_queue_desc_t> ZeCommandQueueDesc;
+  createQueueDesc(QueueIndex, *QueueGroupOrdinal, ZeCommandQueueDesc);
 
   auto ZeResult = ZE_CALL_NOCHECK(
       zeCommandQueueCreate, (Queue->Context->ZeContext, Queue->Device->ZeDevice,
                              &ZeCommandQueueDesc, &ZeQueue));
   if (ZeResult) {
     die("[L0] getZeQueue: failed to create queue");
   }
+
   return ZeQueue;
 }
 
+// This function will return one of possibly multiple available
+// immediate commandlists associated with this Queue.
+pi_command_list_ptr_t &
+_pi_queue::pi_queue_group_t::getImmCmdList(bool UseCopyEngine) {
+
+  uint32_t QueueIndex, QueueOrdinal;
+  auto Index = Queue->getQueueGroup(UseCopyEngine)
+                   .getQueueIndex(&QueueOrdinal, &QueueIndex);
+
+  if (ImmCmdLists[Index] != Queue->CommandListMap.end())
+    return ImmCmdLists[Index];
+
+  ZeStruct<ze_command_queue_desc_t> ZeCommandQueueDesc;
+  createQueueDesc(QueueIndex, QueueOrdinal, ZeCommandQueueDesc);
+
+  ze_command_list_handle_t ZeCommandList;
+  ZE_CALL_NOCHECK(zeCommandListCreateImmediate,
+                  (Queue->Context->ZeContext, Queue->Device->ZeDevice,
+                   &ZeCommandQueueDesc, &ZeCommandList));
+  ImmCmdLists[Index] =
+      Queue->CommandListMap
+          .insert(std::pair<ze_command_list_handle_t, pi_command_list_info_t>{
+              ZeCommandList, {nullptr, false, nullptr, QueueOrdinal}})
+          .first;
+  // Add this commandlist to the cache so it can be destroyed as part of
+  // QueueRelease
+  auto QueueType = Type;
+  auto &ZeCommandListCache =
+      QueueType != queue_type::Compute
+          ? Queue->Context->ZeComputeCommandListCache[Queue->Device->ZeDevice]
+          : Queue->Context->ZeCopyCommandListCache[Queue->Device->ZeDevice];
+  std::lock_guard<std::mutex> lock(Queue->Context->ZeCommandListCacheMutex);
+  ZeCommandListCache.push_back(ZeCommandList);
+
+  return ImmCmdLists[Index];
+}
+
 pi_result _pi_queue::executeOpenCommandListWithEvent(pi_event Event) {
+  if (UseImmediateCommandLists)
+    // When using immediate commandlists there are no open commandlists.
```

Comment:
can we add assertions to check that?

Reply:
The function executeOpenCommandListWithEvent is called from two places, and will be called for both std and imm cmdlists. So an assert is not appropriate. Instead, when in imm cmdlist mode, no further action is needed in this function.

Reply:
I meant to have an assert for "there are no open commandlists". Why cannot these be added?

Reply:
I don't think there is an API for checking whether a commandlist is "open".
For std cmdlists, until they are closed with zeCommandListClose, they are open.
For imm cmdlists the comcept of "open" does not exist because issue is immediate.

Reply:
> I don't think there is an API for checking whether a commandlist is "open".

It's `_pi_queue::hasOpenCommandList`

Reply:
OK, added PI_ASSERT.